File: code/dart/sealed_classes.dart
// A sealed class fixes its set of subtypes, so a switch over it can be
// checked for completeness at compile time.

sealed class Shape {
  const Shape();
}

class Circle extends Shape {
  const Circle(this.radius);
  final double radius;
}

class Rectangle extends Shape {
  const Rectangle(this.width, this.height);
  final double width;
  final double height;
}

class Triangle extends Shape {
  const Triangle(this.base, this.height);
  final double base;
  final double height;
}

// No default branch: adding a fourth Shape would break the build here,
// which is the point.
double area(Shape shape) => switch (shape) {
      Circle(:final radius) => 3.14159265 * radius * radius,
      Rectangle(:final width, :final height) => width * height,
      Triangle(:final base, :final height) => base * height / 2,
    };

String describe(Shape shape) => switch (shape) {
      Rectangle(:final width, :final height) when width == height => 'a square',
      Rectangle() => 'a rectangle',
      Circle(radius: > 10) => 'a large circle',
      Circle() => 'a circle',
      Triangle() => 'a triangle',
    };

sealed class Result<T> {
  const Result();
}

class Ok<T> extends Result<T> {
  const Ok(this.value);
  final T value;
}

class Err<T> extends Result<T> {
  const Err(this.message);
  final String message;
}

Result<int> parseZone(String raw) {
  final zone = int.tryParse(raw);
  if (zone == null) {
    return Err('"$raw" is not a number');
  }
  return zone < 1 || zone > 6 ? Err('zone $zone is outside 1-6') : Ok(zone);
}

void main() {
  const shapes = <Shape>[Circle(2), Rectangle(4, 4), Triangle(6, 2.5), Circle(12)];

  for (final shape in shapes) {
    print('${describe(shape).padRight(16)} area ${area(shape).toStringAsFixed(2)}');
  }

  for (final raw in ['3', '9', 'east']) {
    final message = switch (parseZone(raw)) {
      Ok(:final value) => 'zone $value',
      Err(:final message) => 'rejected: $message',
    };
    print('${raw.padRight(6)} $message');
  }
}

File: code/kotlin/SealedClasses.kt
// A sealed hierarchy has a fixed set of subtypes, so `when` over it needs
// no else branch and is checked for completeness.

import kotlin.math.PI

sealed interface Shape {
    val area: Double
}

data class Circle(val radius: Double) : Shape {
    override val area get() = PI * radius * radius
}

data class Rectangle(val width: Double, val height: Double) : Shape {
    override val area get() = width * height
}

data class Triangle(val base: Double, val height: Double) : Shape {
    override val area get() = base * height / 2
}

// A sealed class with a type parameter, the usual Result shape.
sealed class Outcome<out T> {
    data class Ok<T>(val value: T) : Outcome<T>()
    data class Err(val message: String) : Outcome<Nothing>()
}

fun describe(shape: Shape): String = when (shape) {
    is Rectangle -> if (shape.width == shape.height) "a square" else "a rectangle"
    is Circle -> if (shape.radius > 10) "a large circle" else "a circle"
    is Triangle -> "a triangle"
}

fun parseZone(raw: String): Outcome<Int> {
    val zone = raw.toIntOrNull() ?: return Outcome.Err("\"$raw\" is not a number")
    return if (zone in 1..6) Outcome.Ok(zone) else Outcome.Err("zone $zone is outside 1-6")
}

fun main() {
    val shapes = listOf(Circle(2.0), Rectangle(4.0, 4.0), Triangle(6.0, 2.5), Circle(12.0))

    for (shape in shapes) {
        println("%-16s area %7.2f".format(describe(shape), shape.area))
    }
    println("total %.2f".format(shapes.sumOf { it.area }))

    for (raw in listOf("3", "9", "east")) {
        val message = when (val outcome = parseZone(raw)) {
            is Outcome.Ok -> "zone ${outcome.value}"
            is Outcome.Err -> "rejected: ${outcome.message}"
        }
        println("${raw.padEnd(6)} $message")
    }
}
